### PR TITLE
🪆 Introduction to Matryoshka Embedding Models

### PR DESCRIPTION
**Pull request by tomaarsen**
Hello!

## Pull Request overview
* Add Matryoshka Embedding models blogpost

## Details
This blogpost introduces Matryoshka embeddings, a style of embeddings that has been integrated in Sentence Transformers as of the upcoming 2.4.0 release (on Friday). I intend to release this blogpost alongside the ST release. 

## TODO
- [ ] Verify that the "code URLs" correctly appear as clickable links, otherwise replace them with `<a href="..."><code>...</code></a>`
- [ ] Verify that the embedded demo works.
- [x] Add `Results` section: benchmark results for shrinking embeddings of Matryoshka vs Non-Matryoshka model (still training 2 fresh models for this now)
- [ ] (Hopefully) Transformers.js inference? cc @Xenova
- [ ] Await Sentence Transformers v2.4.0 release on Friday.

cc: @xenova I'm certainly open to other suggestions that you might have! E.g. to showcase Transformers.js as well perhaps.
cc: @osanseviero as you proposed the idea of a blogpost originally.
cc: @philschmid as you expressed an interest in the ST release & subsequent announcement.

- Tom Aarsen

### CODE REVIEW

File: matryoshka.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+## Results
+
+
+## Demo
+
+In this demo, you can dynamically shrink the output dimensions of the [`nomic-ai/nomic-embed-text-v1.5`](https://huggingface.co/nomic-ai/nomic-embed-text-v1.5) Matryoshka embedding model and observe how it affects the retrieval performance. All of the embeddings are computed in the browser using [🤗 Transformers.js](https://github.com/xenova/transformers.js).
+
+<script type="module" src="https://gradio.s3-us-west-2.amazonaws.com/3.32.0/gradio.js"></script>
+<gradio-app theme_mode="light" space="Xenova/adaptive-retrieval-web"></gradio-app>
```
[the remark this answers is not in the document]
Comment (tomaarsen) [reply matryoshka.md:161]: Oh, perfect!

File: matryoshka.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+In practice, this creates an incentive for the model to frontload the most important information in the start of an embedding, such that it will be retained if the embedding is shrunk. 
+
+### In Sentence Transformers
+[Sentence Tranformers](https://sbert.net) is a commonly used framework to train embedding models, and it recently implemented support for Matryoshka models. Training a Matryoshka embedding model using Sentence Transformers is quite elementary: rather than applying some loss function on only the full-size embeddings, we also apply that same loss function on truncated portions of the embeddings.
+
+For example, if a model has an embedding dimension of 768 by default, it can now be trained on 768, 512, 256, 128 and 64. Each of these losses will be added together, optionally with some weight:
```
Comment (xenova) [inline matryoshka.md:47]:
```suggestion
For example, if a model has an original embedding dimension of 768, it can now be trained on 768, 512, 256, 128 and 64. Each of these losses will be added together, optionally with some weight:
```

File: matryoshka.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+loss = MatryoshkaLoss(
+    model=model,
+    loss=base_loss,
+    matryoshka_dims=[768, 512, 256, 128, 64],
+    matryoshka_weight=[1, 1, 1, 1, 1],
+)
+
+model.fit(
+    train_objectives=[(train_dataset, loss)],
+    ...,
```
Comment (xenova) [inline matryoshka.md:65]: I would recommend either (1) defining the list of additional parameters, or (2) linking to where the rest are. Many readers will copy-paste, and without a working snippet, it may annoy them.

File: matryoshka.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+* **[matryoshka_sts.py](https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/matryoshka/matryoshka_sts.py)**: This example uses the CoSENTLoss with MatryoshkaLoss to train an embedding model on the training set of the STSBenchmark dataset. It is an adaptation of the [STS](../sts/README) documentation.
+
+## How do I use 🪆 Matryoshka Embedding models?
+### Theoretically
+In practice, getting embeddings from a Matryoshka embedding model works the same way as with a normal embedding model. The only difference is that after receiving the embeddings, we can optionally truncate them to a smaller dimensionality. Do note that if the embeddings were normalized, then after truncating they will no longer be, so you may want to re-normalize.
+
+Keep in mind that although processing smaller embeddings for downstream tasks (retrieval, clustering, etc.) will be faster, getting the smaller embeddings from the model is just as fast as getting the larger ones.
```
Comment (xenova) [inline matryoshka.md:86]: It might be useful to mention (again) that you can choose to **store** the truncated versions.

File: matryoshka.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+embeddings[..., :matryoshka_dim]  # Shrink the embedding dimensions
+
+# Similarity of the first sentence to the other two:
+similarities = cos_sim(embeddings[0], embeddings[1:])
+print(similarities)
+# => tensor([[0.8428, 0.0873]])
+```
+* Link to the model: [tomaarsen/mpnet-base-nli-matryoshka](https://huggingface.co/tomaarsen/mpnet-base-nli-matryoshka)
+
+Feel free to experiment with  using different values for `matryoshka_dim` and observing how that affects the similarities.
```
Comment (xenova) [inline matryoshka.md:115]:
```suggestion
Feel free to experiment with using different values for `matryoshka_dim` and observing how that affects the similarities.
```

TODO: link to adaptive retrieval demo.

**Review by xenova — COMMENTED** [review]
Will work on images now!

File: matryoshka.md [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+## 🪆 Matryoshka Embeddings
+As research progressed, new state-of-the-art (text) embedding models started producing embeddings with increasingly high output dimensions, i.e. every input text is represented using more values. Although this results in great performance, it does come at a cost of efficiency of downstream tasks such as search or classification.
+
+As a result, [Kusupati et al.](https://arxiv.org/abs/2205.13147) were inspired to create embedding models whose embeddings could reasonably be shrunk without suffering too much on performance.
+
+![matryoshka model](assets/matryoshka/matryoshka_model.png)
+
+These Matryoshka embedding models are trained such that these small truncated embeddings are still useful. In short, Matryoshka embedding models can produce useful embeddings of various dimensions.
+
```
[the remark this answers is not in the document]
Comment (tomaarsen) [reply matryoshka.md:30]: I like the direction, although I would name the section `🪆 Matryoshka Dolls` instead. I'm interested in what you've got planned for the image, too 😄

File: matryoshka.md (hunk 3 above)
Comment (tomaarsen) [reply matryoshka.md:65]: The reason that this isn't a full, complete snippet is because data preprocessing is a pain for Sentence Transformers right now. It's one of the things I'd like to change in v3 (then you can just pass a `Dataset` instance to a `Trainer`).

Right now I do 2: The "References" list below the snippet has the signature/documentation for the `fit` method. All other parameters are "optional, but recommended", so I could also do `model.fit(train_objectives=[(train_dataset, loss)])`, but I don't want to recommend that behaviour.

I also link to complete, ready-to-run scripts below this snippet.

File: matryoshka.md (hunk 4 above)
Comment (tomaarsen) [reply matryoshka.md:88]: I tackled this in 96a5c062c21e56dde44ea8491afea0ea64ad2f26, feel free to review my wording here :)

File: matryoshka.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (xenova) [reply matryoshka.md:30]: Sure thing (edited)

As for the image, I was thinking about possibly creating an animation trying to form an analogy between "discarding smaller dolls" and "truncating the embeddings" while "maintaining most of the information"

File: matryoshka.md [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+* **[matryoshka_sts.py](https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/matryoshka/matryoshka_sts.py)**: This example uses the CoSENTLoss with MatryoshkaLoss to train an embedding model on the training set of the STSBenchmark dataset. It is an adaptation of the [STS](../sts/README) documentation.
+
+## How do I use 🪆 Matryoshka Embedding models?
+### Theoretically
+In practice, getting embeddings from a Matryoshka embedding model works the same way as with a normal embedding model. The only difference is that after receiving the embeddings, we can optionally truncate them to a smaller dimensionality. Do note that if the embeddings were normalized, then after truncating they will no longer be, so you may want to re-normalize.
+
+After truncating, you can either directly apply them for your use cases, or store them such that they can be used later. After all, smaller embeddings in your vector database should result in considerably smaller latencies!
```
Comment (xenova) [inline matryoshka.md:86]:
```suggestion
After truncating, you can either directly apply them for your use cases, or store them such that they can be used later. After all, smaller embeddings in your vector database should result in considerably faster search!
```
or "considerably lower latencies"

File: matryoshka.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
+---
+title: "🪆 Introduction to Matryoshka Embedding Models"
+thumbnail: /blog/assets/matryoshka/thumbnail.png
+authors:
+- user: tomaarsen
+- user: xenova
+---
+
+# 🪆 Introduction to Matryoshka Embedding Models
```
Comment (osanseviero) [inline matryoshka.md:9]: I think we should have a 3-4 sentence description of the blog post + a ToC before the main content

**Review by osanseviero — COMMENTED** [review]
Very very cool!

File: matryoshka.md [hunk 9]
```diff
@@ -0,0 +1,187 @@
+---
+title: "🪆 Introduction to Matryoshka Embedding Models"
+thumbnail: /blog/assets/matryoshka/thumbnail.png
```
[the remark this answers is not in the document]
Comment (tomaarsen) [reply matryoshka.md:3]: Resolved in 80cb46be7063bf9d9fb98e0c2087e5ef2cf529f0

File: matryoshka.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+![embedding model](assets/matryoshka/embedding_model.png)
+
+You can compute the similarity of the complex object by computing the similarity of the embeddings!
+
+![embedding similarity](assets/matryoshka/embedding_similarity.png)
+
+This has an enormous amount of use cases, and serves as the backbone for recommendation systems, retrieval, one-shot or few-shot learning, outlier detection, similarity search, paraphrase detection, clustering, classification, and much more!
+
+## 🪆 Matryoshka Embeddings
+As research progressed, new state-of-the-art (text) embedding models started producing embeddings with increasingly higher output dimensions, i.e., every input text is represented using more values. Although this improves performance, it comes at the cost of efficiency of downstream tasks such as search or classification.
```
[the remark this answers is not in the document]
Comment (tomaarsen) [reply matryoshka.md:23]: Resolved in 80cb46be7063bf9d9fb98e0c2087e5ef2cf529f0

File: matryoshka.md (hunk 7 above)
Comment (tomaarsen) [reply matryoshka.md:86]: I decided to go with "considerable speedups"

File: matryoshka.md [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+Consequently, [Kusupati et al.](https://arxiv.org/abs/2205.13147) (2022) were inspired to create embedding models whose embeddings could reasonably be shrunk without suffering too much on performance.
+
+![matryoshka model](assets/matryoshka/matryoshka_model.png)
+
+These Matryoshka embedding models are trained such that these small truncated embeddings are still useful. In short, Matryoshka embedding models can produce useful embeddings of various dimensions.
+
+## Why would you use 🪆 Matryoshka Embedding models?
+Such variable-size embedding models can be quite valuable to practitioners, for example:
+1. **Shortlisting and reranking**: Rather than performing your downstream task (e.g., nearest neighbor search) on the full embeddings, you can shrink the embeddings to a smaller size and very efficiently "shortlist" your embeddings. Afterwards, you can process the remaining embeddings using their full dimensionality.
+2. **Trade-offs**: Matryoshka models will allow you to scale your embedding solutions to your desired storage cost, processing speed and performance.
```
[the remark this answers is not in the document]
Comment (tomaarsen) [reply matryoshka.md]: Resolved in 7857297aa23602f640da682b92f4fb4de4553ee3

File: matryoshka.md (hunk 8 above)
Comment (tomaarsen) [reply matryoshka.md:9]: Resolved in fbfae0b3336ca18ece4fa24fdf13a1dc7d12eef8 and cbbd67f8deba49786403f4acb0f2167ceedd5fc3, respectively.

File: matryoshka.md [hunk 12]
```diff
@@ -0,0 +1,231 @@
+---
```
Comment (osanseviero) [inline matryoshka.md:1]: You should remove the old thumbnail from the PR?

File: matryoshka.md [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+
+# Similarity of the first sentence to the other two:
+similarities = cos_sim(embeddings[0], embeddings[1:])
+print(similarities)
+# => tensor([[0.8910, 0.1337]])
+```
+
+* Link to the model: [tomaarsen/mpnet-base-nli-matryoshka](https://huggingface.co/tomaarsen/mpnet-base-nli-matryoshka)
+
+Feel free to experiment with using different values for `matryoshka_dim` and observing how that affects the similarities. You can do so either by running this code locally, on the cloud such as with [Google Colab](https://colab.research.google.com/drive/1YlU_T4GaHWVd-PFJFPsTxB6fNNydx0Gx?usp=sharing), or by checking out the [demo](#demo).
```
Comment (osanseviero) [inline matryoshka.md:159]: I would host the notebook on a Hub repo or in GitHub, drive has caused issues in the past (the Hub will automatically how a `Open in Colab` button)

**Review by osanseviero — APPROVED** [review]
Very nice!

File: matryoshka.md (hunk 12 above)
Comment (tomaarsen) [reply matryoshka.md:1]: Resolved in f1acfd6d5443834eea965bb9253b637cd80456f5

File: matryoshka.md (hunk 13 above)
Comment (tomaarsen) [reply matryoshka.md]: Resolved in 3a6e060d098d9e041480439524cc12c4154ae2e9
